cilium/CFP-033-move-hubble-into-cilium.md

```diff
@@ -0,0 +1,120 @@
+# CFP-033: Move `cilium/hubble` into `cilium/cilium`
+
+**SIG: `@cilium/sig-hubble`**
+
+**Begin Design Discussion:** 2024/04/11
+
+**Cilium Release:** 1.16
+
+**Author:** Michi Mutsuzaki (`@michi-covalent`)
+
+**Credit:**
+
+This proposal is based on offline discussions with:
+
+- Fabian Fischer (`@glrf`)
+- Thomas Graf (`@tgraf`)
+- Robin Hahling (`@rolinh`)
+- Anna Kapuścińska (`@lambdanis`)
+- Tobias Klauser (`@tklauser`)
+- Tam Mach (`@sayboras`)
+- Alexandre Perrin (`@kaworu`)
+- Liz Rice (`@lizrice`)
+- Glib Smaga (`@glibsm`)
+- Joe Stringer (`@joestringer`)
+- Sebastian Wicki (`@gandro`)
+- Chance Zibolski (`@chancez`)
+
+## Summary
+
+Move `cilium/hubble` code into `cilium/cilium` repo to improve the Hubble CLI
+development and release processes.
+
+## Motivation
+
+Currently, Hubble server code lives in `cilium/cilium` repo, while Hubble CLI
+code is in a separate `cilium/hubble` repo. This complicates the Hubble CLI
+development and release processes.
+
+### Development process
+
+Currently, modifying Hubble APIs requires dealing with multiple repos:
+
+- Make API and server-side changes in cilium/cilium repo.
+- Update go.mod to pick up the cilium/cilium changes in cilium/hubble repo, and
+  make corresponding client-side changes.
+- Wait for the next Hubble CLI release.
+- Wait for Renovate to update [download-hubble.sh](https://github.com/cilium/cilium/blob/main/images/cilium/download-hubble.sh)
+  before you can use the new Hubble CLI from the `cilium-agent` container.
+
+### Release process
+
+The current Hubble CLI release process requires:
+
+- Vendor unreleased latest `cilium/cilium` code in `cilium/hubble` repo before a
+  new `cilium/cilium` minor version gets released to pick up new Hubble API
+  changes.
+- Release a new Hubble CLI.
+- Make sure [download-hubble.sh](https://github.com/cilium/cilium/blob/main/images/cilium/download-hubble.sh)
+  gets updated before the corresponding `cilium/cilium` minor version gets
+  released.
+ 
+## Goals
+
+The goal of this proposal is to simplify the Hubble CLI development / release
+processes.
+
+* Enable developers to make Hubble server and CLI changes in a single pull
+  request, and immediately make the new Hubble CLI available in the CI
+  `cilium-agent` images without having to wait for the next Hubble CLI release.
+* Build Hubble CLI release binaries from Cilium release tags to simplify the
+  release process.
+
+## Non-Goals
+
+* Archive or delete the cilium/hubble repo.
```

Comment:
Would it be fair to say that it is a non-goal of this CFP to change the versioning schema of Hubble? 

Because I assume this is probably the potentially biggest point of discussion, how we want to version and maintain the Hubble CLI version after Phase II.

Reply:
actually i was thinking the versioning will change as a part of the CFP, basically use the same version as cilium version. there is a section called `Question: How do you version Hubble CLI releases in Phase 1?` below. let me know what you think 💭

Reply:
Ah, right. I overlooked that part, it implies we just use the Cilium version. However, I prefer to have a section about versioning and supported versions, since I believe this does have user impact that I believe we want to make sure we agree on the the path forward. Namely:

- Hubble CLI versions will now be coupled to Cilium versions, i.e. Hubble CLI will jump from 0.13.x to 1.16.y

What I think is less clear to me:

Also, since we would match the Cilium version, what will the maintained versions of the Hubble CLI be? Currently, Hubble only supports one stable branch, where as Cilium maintains three stable branches. Will we release Hubble CLI artifacts for all three stables branches? Do we only release artifacts for the latest stable branch? If we do the latter (i.e. maintain only latest stable), there is going to be the oddity where the "hubble" binary inside the cilium-agent pod effectively is going to be a maintained build of old-stable, i.e. there will be a Hubble CLI 1.16.20 binary inside the cilium-agent 1.16.20 docker image, where as the latest Hubble CLI release on GitHub will be 1.16.8 (or whatever was the latest when we released Hubble CLI v1.17.0). Not necessary a problem, but something to be aware of that this might confuse users.

Another alternative would be to use separate versions for the Hubble CLI inside the cilium-agent pod, and keep having 0.x.y versions for the release artifacts on cilium/hubble (at least feasible during Phase I)

Again, I'm fine to punt those questions to a separate CFP, but I believe we want to have agreement on that at some point, since I do think there is a non-trival decision space here. IMO, we don't have to change the versioning for Phase I, but we probably should change it for Phase II.

Reply:
yeah let me add a section regarding versioning. i have my thoughts 💭

Reply:
updated the CFP. please take a look 🚀🙏

Reply:
Thanks, I think that answers all my questions :pray: